sdk/monitor/azure-monitor-opentelemetry-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/localstorage/LocalStorageTelemetryPipelineListener.java

```diff
@@ -71,21 +74,83 @@ private void processStatusCode206(TelemetryPipelineRequest request, TelemetryPip
         Set<ResponseError> errors = response.getErrors();
         errors.forEach(error -> logger.verbose("Error in telemetry: {}", error));
         if (!errors.isEmpty()) {
-            List<TelemetryItem> originalTelemetryItems = new ArrayList<>();
-            for (ByteBuffer byteBuffer : request.getByteBuffers()) {
-                originalTelemetryItems.addAll(deserialize(byteBuffer.array()));
-            }
-            List<TelemetryItem> toBePersisted = new ArrayList<>();
+            List<ByteBuffer> originalByteBuffers = request.getByteBuffers();
+            byte[] encodedBytes = convertByteBufferListToByteArray(originalByteBuffers);
+            byte[] decodedBytes = ungzip(encodedBytes); // decode is needed in order to split by newline correctly
+            List<byte[]> decodedByteArrayList = splitBytesByNewline(decodedBytes);
+            List<ByteBuffer> toBePersisted = new ArrayList<>();
             for (ResponseError error : errors) {
                 if (StatusCode.isRetryable(error.getStatusCode())) {
-                    toBePersisted.add(originalTelemetryItems.get(error.getIndex()));
+                    toBePersisted.add(ByteBuffer.wrap(decodedByteArrayList.get(error.getIndex())));
                 }
             }
+
             if (!toBePersisted.isEmpty()) {
                 localFileWriter.writeToDisk(
-                    request.getConnectionString(), serialize(toBePersisted), "Received partial response code 206");
+                    request.getConnectionString(), gzip(toBePersisted), "Received partial response code 206");
+            }
+        }
+    }
+
+    // convert list of byte buffers to byte array
+    private static byte[] convertByteBufferListToByteArray(List<ByteBuffer> byteBuffers) {
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        for (ByteBuffer buffer : byteBuffers) {
+            byte[] arr = new byte[buffer.remaining()];
+            buffer.get(arr);
+            try {
+                baos.write(arr);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        return baos.toByteArray();
+    }
+
+    // gzip and adding newline delimiter are required before persisting to the offline disk for handling 206 status code
+    private static List<ByteBuffer> gzip(List<ByteBuffer> byteBuffers) {
```

Comment:
with above changes, I think this signature may be more descriptive
```suggestion
    private static List<ByteBuffer> writePayload(List<byte[]> telemetryItems) {
```

Reply:
how about gzipAndAddNewLine? the main purpose is to gzip before persisting and adds back newlines. it was only preparing content to be written. it doesn't involve writing yet.